File: bench/ai/tasks/varint/starter.cpp
// Decodes a stream of unsigned LEB128 integers given as hexadecimal on standard input.
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>

static int nibble(char c) { return c <= '9' ? c - '0' : c - 'a' + 10; }

static std::vector<std::uint8_t> from_hex(const std::string& hex) {
  std::vector<std::uint8_t> bytes(hex.size() / 2);
  for (std::size_t i = 0; i < bytes.size(); ++i)
    bytes[i] = static_cast<std::uint8_t>(nibble(hex[2 * i]) * 16 + nibble(hex[2 * i + 1]));
  return bytes;
}

enum class Status { ok, truncated, overflow };

// Decodes the integer that starts at `pos`, advancing `pos` past it.
static Status decode(const std::vector<std::uint8_t>& bytes, std::size_t& pos, std::uint64_t& value) {
  value = 0;
  for (unsigned i = 0;; ++i) {
    std::uint8_t b = bytes[pos];
    if (i == 9 && b > 1) return Status::overflow;
    value |= static_cast<std::uint64_t>(b & 0x7f) << (7 * i);
    pos += 1;
    if (b < 0x80) return Status::ok;
  }
}

int main() {
  std::string hex;
  std::cin >> hex;
  std::vector<std::uint8_t> bytes = from_hex(hex);
  std::size_t pos = 0;
  while (pos < bytes.size()) {
    std::size_t start = pos;
    std::uint64_t value = 0;
    Status status = decode(bytes, pos, value);
    if (status == Status::truncated) {
      std::cout << "err truncated " << start << '\n';
      return 0;
    }
    if (status == Status::overflow) {
      std::cout << "err overflow " << start << '\n';
      return 0;
    }
    std::cout << value << '\n';
  }
  return 0;
}

File: bench/ai/tasks/varint/starter.rs
// Decodes a stream of unsigned LEB128 integers given as hexadecimal on standard input.
use std::io::{self, Read, Write};

fn from_hex(hex: &str) -> Vec<u8> {
    hex.as_bytes()
        .chunks(2)
        .map(|p| u8::from_str_radix(std::str::from_utf8(p).unwrap(), 16).unwrap())
        .collect()
}

enum Status {
    Ok(u64),
    Truncated,
    Overflow,
}

// Decodes the integer that starts at `pos`, advancing `pos` past it.
fn decode(bytes: &[u8], pos: &mut usize) -> Status {
    let mut value = 0u64;
    for i in 0.. {
        let b = bytes[*pos];
        if i == 9 && b > 1 {
            return Status::Overflow;
        }
        value |= u64::from(b & 0x7f) << (7 * i);
        *pos += 1;
        if b < 0x80 {
            return Status::Ok(value);
        }
    }
    unreachable!()
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let bytes = from_hex(input.trim());
    let mut out = io::stdout().lock();
    let mut pos = 0;
    while pos < bytes.len() {
        let start = pos;
        match decode(&bytes, &mut pos) {
            Status::Ok(value) => writeln!(out, "{}", value).unwrap(),
            Status::Truncated => {
                writeln!(out, "err truncated {}", start).unwrap();
                return;
            }
            Status::Overflow => {
                writeln!(out, "err overflow {}", start).unwrap();
                return;
            }
        }
    }
}
